### PR TITLE
6 bernhard dev

### PR DESCRIPTION
**Pull request by Kait0**
Changes I made while the others were on vacation.

* Update maps, fixes holes, add traffic light group information (can be used optionally to control TL during eval)
* Update default parameters to match GIGAFLOW more closely
* Fix phantom braking and match to uniform distribution in the paper
* Add true FP32 support (with TF32 / 19 bit turned off)
* Fix a bunch of smaller bugs around spawning
* Made red light infraction be relative to the vehicle center to avoid problems with virtually making the vehicle bigger during eval
* Added jerk threshold for starting to drive when standing still. Remove a bunch of long tail infractions due to low probabilites making the vehicle creep.
* Move noralization more to -1, 1 window
* Average logging acros GPUs.
* remove constant observation
* Add acceleration clipping
* Visualize ego inputs as well
* Disentangle gradient clipping of actor and critic if they are seperate networks
* Implement adavantage filter leakage option that keep a small percent of random samples to prevent bias. Turned off by default
* Add option to visualize training in HTML (removed old EGL stuff)
* Better training cleanup for multinode
* Add paralizable evals
* Some unit tests.

https://wandb.ai/emerge_/nightly-multi-long/runs/k_scaled_0028_1000
1T run. Model gets around 200k km/infraction, which is a new best.

### CODE REVIEW

File: pufferlib/config/puffer_drive.yaml [hunk 1]
```diff
@@ -95,7 +101,9 @@ env:
   # Controller used by non-vehicle agents. "auto" follows non_sdc_controller unless it is "idm", then it uses "replay".
   # options: "auto", "static", "policy", "replay", "idm"
   non_vehicle_controller: auto
-  # Options: "created_all_valid", "create_only_controlled"
+  # Replay-mode agents flagged mark_as_expert are log-replayed (1) or treated like any other agent (0)
+  replay_expert_agents: 1
```
Comment (vcharraut) [inline pufferlib/config/puffer_drive.yaml]: duplicated config

you should be able to do the same thing with 

  # Controller used by non-SDC vehicles.
  # options: "static", "policy", "replay", "idm"
  non_sdc_controller: policy
  # Controller used by non-vehicle agents. "auto" follows non_sdc_controller unless it is "idm", then it uses "replay".
  # options: "auto", "static", "policy", "replay", "idm"
  non_vehicle_controller: auto

in replay

File: pufferlib/ocean/drive/binding.c [hunk 2]
```diff
@@ -1740,10 +1742,11 @@ static PyObject *my_shared(PyObject *self, PyObject *args, PyObject *kwargs) {
...
         free(agent_counts);
 
-        PyObject *tuple = PyTuple_New(3);
+        PyObject *tuple = PyTuple_New(4);
         PyTuple_SetItem(tuple, 0, agent_offsets);
         PyTuple_SetItem(tuple, 1, map_ids_list);
         PyTuple_SetItem(tuple, 2, PyLong_FromLong(env_count));
+        PyTuple_SetItem(tuple, 3, PyLong_FromLong(env_count));
         return tuple;
     }
```
Comment (vcharraut) [inline pufferlib/ocean/drive/binding.c:1751]: this seems useless

File: pufferlib/ocean/drive/binding.c [hunk 3]
```diff
@@ -1677,10 +1678,11 @@ static PyObject *my_shared(PyObject *self, PyObject *args, PyObject *kwargs) {
...
             PyList_SetItem(agent_offsets, env_count, PyLong_FromLong(offset));
 
-            PyObject *tuple = PyTuple_New(3);
+            PyObject *tuple = PyTuple_New(4);
             PyTuple_SetItem(tuple, 0, agent_offsets);
             PyTuple_SetItem(tuple, 1, map_ids_list);
             PyTuple_SetItem(tuple, 2, PyLong_FromLong(env_count));
+            PyTuple_SetItem(tuple, 3, PyLong_FromLong(env_count));
             return tuple;
         }
```
Comment (vcharraut) [inline pufferlib/ocean/drive/binding.c:1687]: this seems useless

File: pufferlib/ocean/drive/drive.h [hunk 4]
```diff
@@ -447,6 +456,9 @@ static void begin_episode_rng(Drive *env) {
         env->episode_seed = rng_next(&env->seed_stream_rng) >> 1;
     }
     rng_seed(&env->rng_state, env->episode_seed);
+    // Per-episode perturbation rates: uniform in [0, configured max]
+    env->episode_partner_blindness_prob = sample_uniform(&env->rng_state, 0.0f, env->partner_blindness_prob);
+    env->episode_phantom_braking_prob = sample_uniform(&env->rng_state, 0.0f, env->phantom_braking_prob);
```
Comment (vcharraut) [inline pufferlib/ocean/drive/drive.h:461]: why do we want a probability for all agents inside an episode vs having a probality per agent ?

File: pufferlib/ocean/drive/drive.h [hunk 5]
```diff
@@ -854,12 +869,62 @@ static bool route_point_at_distance(
...
+        }
+        if (!constrain_heading) {
+            return true;
+        }
+        if (fabsf(normalize_heading(*out_heading - prev_heading)) <= max_heading_delta_rad) {
+            return true;
+        }
+    }
+    return true;
+}
```
Comment (vcharraut) [inline pufferlib/ocean/drive/drive.h:919]: do you know if without this we actually have a lot of goals that are not respecting the constraint  ?

I have deliberatly skipped this constraint because I assumed the goal sampling for GIGAFLOW paper was only done by doing a random sampling of a lane point within an euclidian distance. But if we follow the lane topology, I expect that we have goals with no shard heading most of time

File: pufferlib/ocean/drive/drive.h [hunk 6]
```diff
@@ -2272,54 +2350,16 @@ static void generate_traffic_light_states(Drive *env) {
...
     }
 }
 
 static bool check_spawn_collision(Drive *env, int num_existing_agents, Agent *tmp_agent) {
-    float min_safe_dist_sq = (tmp_agent->sim_length + 5.0f) * (tmp_agent->sim_length + 5.0f);
+    // Inflate the candidate box so agents keep SPAWN_CLEARANCE_M of gap to every neighbor
+    Agent inflated = *tmp_agent;
+    inflated.sim_length += 2.0f * SPAWN_CLEARANCE_M;
+    inflated.sim_width += 2.0f * SPAWN_CLEARANCE_M;
+    float min_safe_dist_sq = (inflated.sim_length + 5.0f) * (inflated.sim_length + 5.0f);
```
Comment (vcharraut) [inline pufferlib/ocean/drive/drive.h]: we should only need one or another

File: pufferlib/ocean/drive/drive.h [hunk 7]
```diff
@@ -2335,19 +2375,19 @@ static bool check_spawn_collision(Drive *env, int num_existing_agents, Agent *tm
...
 }
 
-static bool check_spawn_offroad(Drive *env, Agent *tmp_agent) {
+static bool check_spawn_offroad(Drive *env, Agent *tmp_agent, float edge_clearance_m) {
     // Increase length and width slightly for spawn offroad check
     Agent scaled = *tmp_agent;
-    scaled.sim_length *= 1.1f;
-    scaled.sim_width *= 1.1f;
+    scaled.sim_length = scaled.sim_length * 1.1f + 2.0f * edge_clearance_m;
+    scaled.sim_width = scaled.sim_width * 1.1f + 2.0f * edge_clearance_m;
```
Comment (vcharraut) [inline pufferlib/ocean/drive/drive.h]: too many magic number

File: pufferlib/ocean/drive/drive.h [hunk 8]
```diff
@@ -2537,15 +2577,122 @@ static bool spawn_agent(Drive *env, int agent_idx, int num_agents) {
...
+        }
+        if (fabsf(other->sim_z - inflated.sim_z) > Z_BUFFER) {
+            continue;
+        }
+        if (check_obb_collision(&inflated, other)) {
+            return false;
+        }
+    }
+    return true;
+}
```
Comment (vcharraut) [inline pufferlib/ocean/drive/drive.h:2689]: Ideally this code should be in 123Drive.
But for direct solution as it is, it would better to do inside the init function, similar to remove_bad_trajectories. Either enhance this function or append a new function after

File: pufferlib/ocean/drive/drive.h (hunk 5 above)
Comment (Kait0) [reply pufferlib/ocean/drive/drive.h:919]: I just tried to match it to the gigaflow implementation. I don't think this made any major difference

File: pufferlib/config/puffer_drive.yaml (hunk 1 above)
Comment (Kait0) [reply pufferlib/config/puffer_drive.yaml]: So I only use replay_expert_agents: 0
for nuPlan and CARLA

Claude claims the 1 branch is necessary for compatibility with WOMD, but I never used WOMD, so idk.
The controllers are category-wide, whereas this flag uses the expert/non-expert flag from the dataset
I would keep it if it doesn't bother.

File: pufferlib/ocean/drive/binding.c (hunk 3 above)
Comment (Kait0) [reply pufferlib/ocean/drive/binding.c:1687]: Slot 3 is maps_consumed, Slot 2 is num_envs. Both are used in the code. We could also set the equivalence elsewhere I guess?

File: pufferlib/ocean/drive/drive.h (hunk 4 above)
Comment (Kait0) [reply pufferlib/ocean/drive/drive.h:461]: To match the gigaflow paper which have an up to 5% number of phantom brakers. That is the total number of phatom brakers in an episode.
The individual phantom braker still has a probability for it to trigger.
We want agents that experience clean episodes and agents that experience crazy epsiodes I think.
Having it per agent would mean that on average all scenarios look the same due to many draws?

File: pufferlib/ocean/drive/drive.h (hunk 6 above)
Comment (Kait0) [reply pufferlib/ocean/drive/drive.h]: There are two checks a cheap distance based one for pre-filter and a more expensive one for bb detection. 
think different sizes makes sense, but I will change the magic number 5.0,

File: pufferlib/ocean/drive/drive.h (hunk 8 above)
Comment (Kait0) [reply pufferlib/ocean/drive/drive.h:2689]: I think spawn point selection, is somewhat user specific not map specific?
Think this is called twice throughout the code so in the init doesn't make so much sense.

File: pufferlib/ocean/drive/drive.h (hunk 7 above)
Comment (Kait0) [reply pufferlib/ocean/drive/drive.h]: gave it a name.

File: pufferlib/ocean/drive/drive.h (hunk 4 above)
Comment (vcharraut) [reply pufferlib/ocean/drive/drive.h:461]: Ok I understand. In my comprehension of the paper, the "Up to 10% of agents" refers to all the agents batch-wise, so not per episode

I have no idea which one is correct

Comment (Kait0) [reply pufferlib/ocean/drive/drive.h:461]: yeah, we can ask eugene when he is back. The uniform distribution I use is also an assumption. Though I think this change did help to reduce the rear end collisions, so I would keep it unless eugene says he has a better solution.

File: pufferlib/config/puffer_drive.yaml (hunk 1 above)
Comment (vcharraut) [reply pufferlib/config/puffer_drive.yaml]: I see, I quoted the wrong config then

You could add a new option to `control_mode` like `control_all_vehicles` that would actually control all valid cars, enforcing controlling the mark_as_expert ones 

Ideally we don't want to create one config variable for every behavior we rather try to merge into existing one when possible

File: pufferlib/config/puffer_drive.yaml [hunk 9]
```diff
@@ -271,19 +296,11 @@ train:
   adv_filter_enabled: true
   adv_filter_ewma_beta: 0.25
   adv_filter_threshold_scale: 0.01
+  # Also keep this random fraction of below-threshold transitions; 0 disables the leak
+  adv_filter_leak_fraction: 0.0
```
Comment (vcharraut) [inline pufferlib/config/puffer_drive.yaml:300]: is this used in training ?

File: pufferlib/config/puffer_drive.yaml [hunk 10]
```diff
@@ -230,6 +248,8 @@ train:
   optimizer: adamw
   anneal_lr: true
   precision: bfloat16
+  # false: pure float32 matmuls/convs (no TensorFloat-32); requires precision float32
+  tf32: true
```
Comment (vcharraut) [inline pufferlib/config/puffer_drive.yaml:252]: i think we want a better naming for this one

File: pufferlib/ocean/drive/drive.h [hunk 11]
```diff
@@ -1483,18 +1557,18 @@ static bool check_agent_on_stop_line(Drive *env, Agent *agent, bool include_yell
...
-    float prev_rear_x = agent->prev_x + rear_offset * agent->prev_cos_heading;
-    float prev_rear_y = agent->prev_y + rear_offset * agent->prev_sin_heading;
+static bool check_red_light_violation(Drive *env, int agent_idx) {
+    Agent *agent = &env->agents[agent_idx];
+    // Vehicle center crosses on red; the laterally extended line cannot be driven
+    // around, and the heading gate exempts exiting/opposing traffic.
+    float center_x = agent->sim_x;
+    float center_y = agent->sim_y;
+    float prev_center_x = agent->prev_x;
+    float prev_center_y = agent->prev_y;
```
Comment (vcharraut) [inline pufferlib/ocean/drive/drive.h:1567]: use directly the variable, no need for aliases

File: pufferlib/ocean/drive/drive.h (hunk 5 above)
Comment (vcharraut) [reply pufferlib/ocean/drive/drive.h:919]: I'm afraid this implementation can favorise closer goals -> the closer the goal is, the more likely it is to respect the 60 degrees contrainst

It would be preferable to move this change to a dedicated PR so we can better emphasize the impact of it

File: pufferlib/config/puffer_drive.yaml (hunk 9 above)
Comment (Kait0) [reply pufferlib/config/puffer_drive.yaml:300]: Currently no. I tried 5% leak but no conclusive improvement. Left it there because I think it is a sensible idea that I might want to revisit later.
